src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,470 @@
+            The amount of XRP the signature is for (not needed if equal to Balance or just
+            closing the line).
+        Signature (optional)
+            Authorization for the balance above, signed by the owner (optional,
+            not needed if closing or owner is performing the transaction). The
+            signature if for the following message: CLM\0 followed by the
+            256-bit channel ID, and a 64-bit integer drops.
+        PublicKey (optional)
+           The public key that made the signature (optional, required if a signature
+           is present)
+        Flags
+            tfCloseChannel
+                Request that the channel be closed
+            tfRenewChannel
+                Request that the channel's expiration be reset. Only the owner
+                may renew a channel.
+
+*/
+
+//------------------------------------------------------------------------------
+
+static
+TER
+closeChannel (
+    std::shared_ptr<SLE> const& slep,
+    ApplyView& view,
+    uint256 const& key,
+    beast::Journal j)
+{
+    AccountID const src = (*slep)[sfAccount];
+    // Remove PayChan from owner directory
+    {
+        auto const page = (*slep)[sfOwnerNode];
+        TER const ter = dirDelete (view, true, page, keylet::ownerDir (src).key,
+            key, false, page == 0, j);
+        if (!isTesSuccess (ter))
+            return ter;
+    }
+
+    // Transfer amount back to owner, decrement owner count
+    auto const sle = view.peek (keylet::account (src));
+    if (!sle)
+        return terNO_ACCOUNT;
+    assert ((*slep)[sfAmount] >= (*slep)[sfBalance]);
+    (*sle)[sfBalance] =
+        (*sle)[sfBalance] + (*slep)[sfAmount] - (*slep)[sfBalance];
+    (*sle)[sfOwnerCount] = (*sle)[sfOwnerCount] - 1;
+    view.update (sle);
+
+    // Remove PayChan from ledger
+    view.erase (slep);
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanCreate::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    if (!isXRP (ctx.tx[sfAmount]) || (ctx.tx[sfAmount] <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (ctx.tx[sfAccount] == ctx.tx[sfDestination])
+        return temMALFORMED;
+
+    return preflight2 (ctx);
+}
+
+TER
+PayChanCreate::doApply()
+{
+    auto const account = ctx_.tx[sfAccount];
+    auto const sle = ctx_.view ().peek (keylet::account (account));
+
+    // Check reserve and funds availability
+    {
+        auto const balance = (*sle)[sfBalance];
+        auto const reserve =
+            ctx_.view ().fees ().accountReserve ((*sle)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+
+        if (balance < reserve + ctx_.tx[sfAmount])
+            return tecUNFUNDED;
+    }
```

Comment:
I believe that this block, as well as the next (so lines 176-201) can be handled in `preclaim`. Is there a reason why we do this in `doApply`?